File: rust_v2/src/domain/orderbook/wire.rs
//! Wire types for orderbook responses (REST + WS).

use crate::shared::{OrderBookId, Side};
use chrono::{DateTime, Utc};
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};

// ─── REST wire types ─────────────────────────────────────────────────────────

/// REST response for a single orderbook.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OrderbookResponse {
    pub id: i32,
    pub market_pubkey: String,
    pub orderbook_id: String,
    pub base_token: String,
    pub quote_token: String,
    pub outcome_index: Option<i16>,
    pub tick_size: i64,
    pub total_bids: i32,
    pub total_asks: i32,
    pub last_trade_price: Option<Decimal>,
    pub last_trade_time: Option<DateTime<Utc>>,
    pub active: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// REST response for multiple orderbooks.
#[derive(Debug, Serialize, Deserialize)]
pub struct OrderbooksResponse {
    pub orderbooks: Vec<OrderbookResponse>,
    pub total: usize,
}

/// REST response for orderbook depth.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OrderbookDepthResponse {
    pub orderbook_id: OrderBookId,
    pub best_bid: Option<Decimal>,
    pub best_ask: Option<Decimal>,
    pub bids: Vec<BookOrder>,
    pub asks: Vec<BookOrder>,
}

/// Decimals metadata for an orderbook.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DecimalsResponse {
    pub orderbook_id: String,
    pub base_decimals: u8,
    pub quote_decimals: u8,
    pub price_decimals: u8,
}

// ─── WS wire types ───────────────────────────────────────────────────────────

/// WS orderbook snapshot or delta.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct OrderBook {
    #[serde(rename = "orderbook_id")]
    pub id: OrderBookId,
    #[serde(default)]
    pub is_snapshot: bool,
    #[serde(default)]
    pub seq: u32,
    #[serde(default = "Vec::new")]
    pub bids: Vec<BookOrder>,
    #[serde(default = "Vec::new")]
    pub asks: Vec<BookOrder>,
}

/// A single price level in the order book.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct BookOrder {
    pub side: Side,
    pub size: Decimal,
    pub price: Decimal,
}

/// WS ticker data.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WsTickerData {
    pub orderbook_id: OrderBookId,
    pub best_bid: Option<Decimal>,
    pub best_ask: Option<Decimal>,
    pub mid_price: Option<Decimal>,
}

File: python/src/domain/orderbook/wire.py
"""Orderbook wire types."""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class PriceLevel:
    price: str
    size: str

    @staticmethod
    def from_dict(d: dict) -> "PriceLevel":
        return PriceLevel(price=d.get("price", "0"), size=d.get("size", "0"))

    @staticmethod
    def from_list(lst: list) -> "PriceLevel":
        return PriceLevel(price=str(lst[0]) if lst else "0", size=str(lst[1]) if len(lst) > 1 else "0")


@dataclass
class OrderbookDepthResponse:
    bids: list[PriceLevel] = field(default_factory=list)
    asks: list[PriceLevel] = field(default_factory=list)
    orderbook_id: Optional[str] = None

    @staticmethod
    def from_dict(d: dict) -> "OrderbookDepthResponse":
        return OrderbookDepthResponse(
            bids=[PriceLevel.from_dict(b) if isinstance(b, dict) else PriceLevel.from_list(b) for b in d.get("bids", [])],
            asks=[PriceLevel.from_dict(a) if isinstance(a, dict) else PriceLevel.from_list(a) for a in d.get("asks", [])],
            orderbook_id=d.get("orderbook_id"),
        )


@dataclass
class DecimalsResponse:
    base_decimals: int = 6
    quote_decimals: int = 6
    price_decimals: int = 6

    @staticmethod
    def from_dict(d: dict) -> "DecimalsResponse":
        return DecimalsResponse(
            base_decimals=d.get("base_decimals", 6),
            quote_decimals=d.get("quote_decimals", 6),
            price_decimals=d.get("price_decimals", 6),
        )


@dataclass
class WsTickerData:
    orderbook_id: str
    best_bid: Optional[str] = None
    best_ask: Optional[str] = None
    mid_price: Optional[str] = None
    last_trade_price: Optional[str] = None

    @staticmethod
    def from_dict(d: dict) -> "WsTickerData":
        return WsTickerData(
            orderbook_id=d.get("orderbook_id", ""),
            best_bid=d.get("best_bid"),
            best_ask=d.get("best_ask"),
            mid_price=d.get("mid_price"),
            last_trade_price=d.get("last_trade_price"),
        )
